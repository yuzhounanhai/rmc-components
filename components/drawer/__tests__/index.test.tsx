import React from 'react';
import { act, Simulate } from 'react-dom/test-utils';
import BaseDrawer from '../drawer';
import Drawer from '..';
import { mount } from 'enzyme';
import FadeIn from '../../fade/fadeIn';
import SlideIn from '../../slide/slideIn';
import baseTest from '../../../tests/common/baseTest';
import { sleep } from '../../../tests/common/utils';
import baseReactDOMTest, { $$ } from '../../../tests/common/baseReactDOMTest';
import { defaultPrefixCls } from '../../_config/dict';

describe('Drawer', () => {
  baseReactDOMTest();

  baseTest(
    <BaseDrawer>
      <div>content</div>
    </BaseDrawer>
  );

  baseTest(
    <BaseDrawer
      direction="â†“"
    >
      <div>content</div>
    </BaseDrawer>
  );

  baseTest(
    <BaseDrawer
      onShow={jest.fn()}
      onHide={jest.fn()}
      onCancel={jest.fn()}
      visible={false}
    >
      <div>content</div>
    </BaseDrawer>
  );

  baseTest(
    <BaseDrawer
      onCancel={jest.fn()}
      visible
    >
      {
        (onTriggerCancel) => (
          <div
            onClick={onTriggerCancel}
          >
            close
          </div>
        )
      }
    </BaseDrawer>
  );

  it('should trigger onShow/onHide callback when <Drawer /> visible status changed', async () => {
    const onHideCb = jest.fn();
    const onShowCb = jest.fn();
    const wrapper = mount(
      <Drawer
        onShow={onShowCb}
        onHide={onHideCb}
        visible={false}
      >
        <div>content</div>
      </Drawer>
    );
    await act(async () => {
      await sleep(100);
      wrapper.update();
    });
    expect(onShowCb).not.toBeCalled();
    expect(onHideCb).not.toBeCalled();
    wrapper.setProps({
      visible: true,
    });
    wrapper.update();
    wrapper.find(SlideIn).simulate('transitionend');
    wrapper.find(FadeIn).simulate('transitionend');
    await sleep(50);
    expect(onShowCb).toBeCalledTimes(1);
    expect(onHideCb).toBeCalledTimes(0);
    wrapper.setProps({
      visible: false,
    });
    wrapper.update();
    wrapper.find(SlideIn).simulate('transitionend');
    wrapper.find(FadeIn).simulate('transitionend');
    await sleep(50);
    expect(onShowCb).toBeCalledTimes(1);
    expect(onHideCb).toBeCalledTimes(1);
    const wrapper1 = mount(
      <Drawer
        mask={false}
        visible={false}
        onShow={onShowCb}
        onHide={onHideCb}
      >
        <div>content</div>
      </Drawer>
    );
    await act(async () => {
      await sleep(100);
      wrapper1.update();
    });
    wrapper1.setProps({
      visible: true,
    });
    wrapper1.update();
    wrapper1.find(SlideIn).simulate('transitionend');
    await sleep(50);
    expect(onShowCb).toBeCalledTimes(2);
    expect(onHideCb).toBeCalledTimes(1);
    wrapper1.setProps({
      visible: false,
    });
    wrapper1.update();
    wrapper1.find(SlideIn).simulate('transitionend');
    await sleep(50);
    expect(onShowCb).toBeCalledTimes(2);
    expect(onHideCb).toBeCalledTimes(2);
  });

  it('should call onCancel when trigger close action.', async () => {
    const onCancelCb = jest.fn();
    const wrapper = mount(
      <Drawer
        onCancel={onCancelCb}
        maskClosable
        prefixCls="rc"
      >
        {
          (onTriggerClose) => (
            <div
              className="btn-close"
              onClick={onTriggerClose}
            >
              close
            </div>
          )
        }
      </Drawer>
    );
    await act(async () => {
      await sleep(100);
      wrapper.update();
    });
    wrapper.find('.rc-drawer-mask').simulate('click');
    expect(onCancelCb).toBeCalledTimes(1);
    wrapper.find('.btn-close').simulate('click');
    expect(onCancelCb).toBeCalledTimes(2);
    wrapper.setProps({
      maskClosable: false,
    });
    wrapper.update();
    wrapper.find('.rc-drawer-mask').simulate('click');
    expect(onCancelCb).toBeCalledTimes(2);
  });

  it('should not render mask when mask set FALSE.', async () => {
    const wrapper = mount(
      <Drawer
        mask={false}
        prefixCls="rc"
      >
        <div>content</div>
      </Drawer>
    );
    expect(wrapper.find('.rc-drawer.mask')).toHaveLength(0);
  });

  it('Drawer.createDrawer should be defined', () => {
    expect(Drawer.createDrawer).toBeDefined();
  });

  it('Drawer should run correctly whick created by "createDrawer" without "onCancel" parameter.', async () => {
    const onHideCb = jest.fn();
    Drawer.createDrawer({
      content: 'content',
      onHide: onHideCb,
      className: 'test-drawer',
    });
    expect($$('.test-drawer')).not.toBeNull();
    await act(async () => {
      await sleep(50);
    });
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-slide`));
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-fade`));
    Simulate.click($$(`.${defaultPrefixCls}-drawer-mask`));
    expect(onHideCb).not.toBeCalled();
    await act(async () => {
      await sleep(10);
    });
    expect(() => {
      Simulate.click($$(`.${defaultPrefixCls}-drawer-mask`));
    }).not.toThrow();
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-slide`));
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-fade`));
    expect(onHideCb).toBeCalled();
    expect($$('.test-drawer')).toBeNull();
  });

  it('Drawer should run correctly whick created by "createDrawer" with "onCancel" parameter.', async () => {
    const onHideCb = jest.fn();
    const onCancelCb = jest.fn();
    Drawer.createDrawer({
      content: 'content',
      onHide: onHideCb,
      onCancel: (close) => {
        close();
        onCancelCb();
      },
      className: 'test-drawer',
    });
    expect($$('.test-drawer')).not.toBeNull();
    await act(async () => {
      await sleep(50);
    });
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-slide`));
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-fade`));
    Simulate.click($$(`.${defaultPrefixCls}-drawer-mask`));
    expect(onCancelCb).toBeCalled();
    expect(onHideCb).not.toBeCalled();
    await act(async () => {
      await sleep(10);
    });
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-slide`));
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-fade`));
    expect(onHideCb).toBeCalled();
    expect($$('.test-drawer')).toBeNull();
  });
});