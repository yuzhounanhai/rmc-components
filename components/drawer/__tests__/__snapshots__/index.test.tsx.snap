// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer component snapshots tests 1`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "class": "rmcc-drawer",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "class": "rmcc-drawer-mask rmcc-fade rmcc-fade-op-0",
          "style": "transition-property:opacity;transition-duration:0.5s;transition-timing-function:ease",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Node {
          "attribs": Object {
            "class": "rmcc-drawer-content rmcc-drawer-d-up rmcc-slide rmcc-slide-begin-up rmcc-slide-op-0",
            "style": "transition-property:transform,opacity;transition-duration:0.5s;transition-timing-function:ease",
          },
          "children": Array [
            Node {
              "attribs": Object {
                "class": "rmcc-drawer-box",
              },
              "children": Array [
                Node {
                  "attribs": Object {},
                  "children": Array [
                    Node {
                      "data": "content",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
      Node {
        "attribs": Object {
          "class": "rmcc-drawer-content rmcc-drawer-d-up rmcc-slide rmcc-slide-begin-up rmcc-slide-op-0",
          "style": "transition-property:transform,opacity;transition-duration:0.5s;transition-timing-function:ease",
        },
        "children": Array [
          Node {
            "attribs": Object {
              "class": "rmcc-drawer-box",
            },
            "children": Array [
              Node {
                "attribs": Object {},
                "children": Array [
                  Node {
                    "data": "content",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Node {
          "attribs": Object {
            "class": "rmcc-drawer-mask rmcc-fade rmcc-fade-op-0",
            "style": "transition-property:opacity;transition-duration:0.5s;transition-timing-function:ease",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Drawer component snapshots tests 2`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "class": "rmcc-drawer",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "class": "rmcc-drawer-mask rmcc-fade rmcc-fade-op-0",
          "style": "transition-property:opacity;transition-duration:0.5s;transition-timing-function:ease",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Node {
          "attribs": Object {
            "class": "rmcc-drawer-content rmcc-drawer-d-down rmcc-slide rmcc-slide-begin-down rmcc-slide-op-0",
            "style": "transition-property:transform,opacity;transition-duration:0.5s;transition-timing-function:ease",
          },
          "children": Array [
            Node {
              "attribs": Object {
                "class": "rmcc-drawer-box",
              },
              "children": Array [
                Node {
                  "attribs": Object {},
                  "children": Array [
                    Node {
                      "data": "content",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
      Node {
        "attribs": Object {
          "class": "rmcc-drawer-content rmcc-drawer-d-down rmcc-slide rmcc-slide-begin-down rmcc-slide-op-0",
          "style": "transition-property:transform,opacity;transition-duration:0.5s;transition-timing-function:ease",
        },
        "children": Array [
          Node {
            "attribs": Object {
              "class": "rmcc-drawer-box",
            },
            "children": Array [
              Node {
                "attribs": Object {},
                "children": Array [
                  Node {
                    "data": "content",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Node {
          "attribs": Object {
            "class": "rmcc-drawer-mask rmcc-fade rmcc-fade-op-0",
            "style": "transition-property:opacity;transition-duration:0.5s;transition-timing-function:ease",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Drawer component snapshots tests 3`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "class": "rmcc-drawer",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "class": "rmcc-drawer-mask rmcc-fade rmcc-fade-op-0",
          "style": "transition-property:opacity;transition-duration:0.5s;transition-timing-function:ease",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Node {
          "attribs": Object {
            "class": "rmcc-drawer-content rmcc-drawer-d-up rmcc-slide rmcc-slide-begin-up rmcc-slide-op-0",
            "style": "transition-property:transform,opacity;transition-duration:0.5s;transition-timing-function:ease",
          },
          "children": Array [
            Node {
              "attribs": Object {
                "class": "rmcc-drawer-box",
              },
              "children": Array [
                Node {
                  "attribs": Object {},
                  "children": Array [
                    Node {
                      "data": "content",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
      Node {
        "attribs": Object {
          "class": "rmcc-drawer-content rmcc-drawer-d-up rmcc-slide rmcc-slide-begin-up rmcc-slide-op-0",
          "style": "transition-property:transform,opacity;transition-duration:0.5s;transition-timing-function:ease",
        },
        "children": Array [
          Node {
            "attribs": Object {
              "class": "rmcc-drawer-box",
            },
            "children": Array [
              Node {
                "attribs": Object {},
                "children": Array [
                  Node {
                    "data": "content",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Node {
          "attribs": Object {
            "class": "rmcc-drawer-mask rmcc-fade rmcc-fade-op-0",
            "style": "transition-property:opacity;transition-duration:0.5s;transition-timing-function:ease",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Drawer component snapshots tests 4`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "class": "rmcc-drawer",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "class": "rmcc-drawer-mask rmcc-fade rmcc-fade-op-0",
          "style": "transition-property:opacity;transition-duration:0.5s;transition-timing-function:ease",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Node {
          "attribs": Object {
            "class": "rmcc-drawer-content rmcc-drawer-d-up rmcc-slide rmcc-slide-begin-up rmcc-slide-op-0",
            "style": "transition-property:transform,opacity;transition-duration:0.5s;transition-timing-function:ease",
          },
          "children": Array [
            Node {
              "attribs": Object {
                "class": "rmcc-drawer-box",
              },
              "children": Array [
                Node {
                  "attribs": Object {},
                  "children": Array [
                    Node {
                      "data": "close",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
      Node {
        "attribs": Object {
          "class": "rmcc-drawer-content rmcc-drawer-d-up rmcc-slide rmcc-slide-begin-up rmcc-slide-op-0",
          "style": "transition-property:transform,opacity;transition-duration:0.5s;transition-timing-function:ease",
        },
        "children": Array [
          Node {
            "attribs": Object {
              "class": "rmcc-drawer-box",
            },
            "children": Array [
              Node {
                "attribs": Object {},
                "children": Array [
                  Node {
                    "data": "close",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Node {
          "attribs": Object {
            "class": "rmcc-drawer-mask rmcc-fade rmcc-fade-op-0",
            "style": "transition-property:opacity;transition-duration:0.5s;transition-timing-function:ease",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
